File: 0143-reorder-list/0143-reorder-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:

    void reorderList(ListNode* head) {

        ListNode* fast = head->next;
        ListNode* slow = head;

        while (fast and fast->next) // finding middle
        {
            slow = slow->next;
            fast = fast->next->next;
        }

        ListNode* second = slow->next;
        ListNode* prev = slow->next = NULL;
        while (second) // reversing
        {
            ListNode* next = second->next;
            second->next = prev;
            prev = second;
            second = next;
        }
        second = prev; // second - beginning of reversed list
        ListNode* first = head; // first - beginning of list
        
        while(second)
        {
            ListNode* next1 = first->next;
            ListNode* next2 = second->next;

            first->next = second;
            second->next = next1;
            first = next1;
            second = next2;
        }
        
    }
};

File: 0143-reorder-list/0143-reorder-list.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class Solution {
    public void ReorderList(ListNode head) {
        ListNode slow = head, fast = head.next;

        while (fast != null && fast.next != null)
        {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode l2 = slow.next;
        ListNode prev = slow.next = null;
        while (l2 != null) // reversing
        {
            ListNode next = l2.next;
            l2.next = prev;
            prev = l2;
            l2 = next;
        }

        l2 = prev;
        ListNode first = head;

        while (l2 != null)
        {
            ListNode next1 = first.next;
            ListNode next2 = l2.next;

            first.next = l2;
            l2.next = next1;
            first = next1;
            l2 = next2;
        }
        
    }

}
